manager/orchestrator/replicated/task_reaper_test.go

```diff
@@ -569,3 +567,149 @@ func TestServiceRemoveDeadTasks(t *testing.T) {
+	defer cancel()
+
+	service1 := &api.Service{
+		ID: "id1",
+		Spec: api.ServiceSpec{
+			Annotations: api.Annotations{
+				Name: "name1",
+			},
+			Mode: &api.ServiceSpec_Replicated{
+				Replicated: &api.ReplicatedService{
+					Replicas: 1,
+				},
+			},
+			Task: api.TaskSpec{
+				Restart: &api.RestartPolicy{
+					// Turn off restart to get an accurate count on tasks.
+					Condition: api.RestartOnNone,
+					Delay:     gogotypes.DurationProto(0),
+				},
+			},
+		},
+	}
+
+	// Create a service with one replica specified before the orchestrator is
+	// started. This should result in two tasks when the orchestrator
+	// starts up.
+	err := s.Update(func(tx store.Tx) error {
+		assert.NoError(t, store.CreateService(tx, service1))
+		return nil
+	})
+	assert.NoError(t, err)
+
+	// Start the orchestrator.
+	go func() {
+		assert.NoError(t, orchestrator.Run(ctx))
+	}()
+
+	observedTask1 := testutils.WatchTaskCreate(t, watch)
+	assert.Equal(t, api.TaskStateNew, observedTask1.Status.State)
+	assert.Equal(t, observedTask1.ServiceAnnotations.Name, "name1")
+
+	// Set the task state to PENDING to simulate allocation.
+	updatedTask1 := observedTask1.Copy()
+	updatedTask1.Status.State = api.TaskStatePending
+	updatedTask1.ServiceAnnotations = api.Annotations{Name: "original"}
+	err = s.Update(func(tx store.Tx) error {
+		assert.NoError(t, store.UpdateTask(tx, updatedTask1))
+		return nil
+	})
+	require.NoError(t, err)
+
+	testutils.Expect(t, watch, state.EventCommit{})
+	testutils.Expect(t, watch, api.EventUpdateTask{})
+	testutils.Expect(t, watch, state.EventCommit{})
+
+	// Loop twice:
+	// 1. first time simulates existence of a task in state PENDING
+	//	and desired state SHUTDOWN before the task reaper starts up.
+	// 2. second time simulates task reaper receiving a task update event
+	//	for a task in state PENDING and desired state SHUTDOWN and the reaper
+	//	taking action to clean that task up.
+	for i := 0; i < 2; i++ {
+		// Force Update the service.
+		service1.Spec.Task.ForceUpdate++
+		// This should shutdown the previous task and create a new one.
+		err = s.Update(func(tx store.Tx) error {
+			assert.NoError(t, store.UpdateService(tx, service1))
+			return nil
+		})
+		testutils.Expect(t, watch, api.EventUpdateService{})
+		testutils.Expect(t, watch, state.EventCommit{})
+
+		// New task should be created and old task marked for SHUTDOWN.
+		observedTask1 = testutils.WatchTaskCreate(t, watch)
+		assert.Equal(t, api.TaskStateNew, observedTask1.Status.State)
+		assert.Equal(t, observedTask1.ServiceAnnotations.Name, "name1")
+
+		observedTask3 := testutils.WatchTaskUpdate(t, watch)
+		assert.Equal(t, api.TaskStateShutdown, observedTask3.DesiredState)
+		assert.Equal(t, "original", observedTask3.ServiceAnnotations.Name)
+
+		testutils.Expect(t, watch, state.EventCommit{})
+
+		// Set the task state for the restarted task PENDING to simulate allocation.
```

Comment:
Non-blocking:  To simulate what actually happens, should this be set to pending before the force update, so that the task is already in pending state when it is shut down?  If not, why must it occur here, instead of before the force update?